From ff690fbcebc67210cd72f9018d9d5a4090b2199d Mon Sep 17 00:00:00 2001
From: Benjamin Coddington <bcodding@redhat.com>
Date: Tue, 6 Jan 2015 16:46:10 -0500
Subject: [fs] nfsd4: fix xdr4 inclusion of escaped char

Message-id: <ea70b66f5d637ca26f88fb446bf719398cb8c935.1420562070.git.bcodding@redhat.com>
Patchwork-id: 102495
O-Subject: [RHEL7.1 PATCH fs BZ1164055 1/2] nfsd4: fix xdr4 inclusion of escaped char
Bugzilla: 1164055
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

Fix a bug where nfsd4_encode_components_esc() includes the esc_end char as
an additional string encoding.

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1164055
Upstream Id: 5a64e56976f1ba98743e1678c0029a98e9034c81
Signed-off-by: Benjamin Coddington <bcodding@redhat.com>
Cc: stable@vger.kernel.org
Fixes: e7a0444aef4a "nfsd: add IPv6 addr escaping to fs_location hosts"
Signed-off-by: J. Bruce Fields <bfields@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4xdr.c b/fs/nfsd/nfs4xdr.c
index b01f6e1..d53d792 100644
--- a/fs/nfsd/nfs4xdr.c
+++ b/fs/nfsd/nfs4xdr.c
@@ -1764,6 +1764,9 @@ static __be32 nfsd4_encode_components_esc(struct xdr_stream *xdr, char sep,
   }
   else
    end++;
+  if (found_esc)
+   end = next;
+
   str = end;
  }
  pathlen = htonl(xdr->buf->len - pathlen_offset);
-- 
1.7.1